docs: add guide for signal inputs (#54316)
angular/angular: aio/content/guide/signal-inputs.md

# Signal inputs

> Signal inputs are a reactive alternative to decorator-based `@Input()`.

Reviewer: This still feels like we're leading with the comparison with the existing `@Input`. We might want to jump directly into the text from the Overview section and move the alert about dev preview lower in that section.

<div class="alert is-helpful">

Signal inputs are currently in [developer preview](/guide/releases#developer-preview).

</div>

## Overview

Signal inputs allow values to be bound from parent components.
Those values are exposed using a `Signal` and might change during the lifecycle of your component.

> Angular supports two variants of inputs:

Author: (as discussed offline- will follow-up on this with you)

**Optional inputs**
Inputs are optional by default, unless you use `input.required`.
You can specify an explicit initial value, or Angular will use `undefined` implicitly.

**Required inputs**
Required inputs always have a value of the given input type.

> They are declared using the `input.required` function.

Author: Interesting. This is clearly subjective, but it does feel like I'd prefer a good intro for the two possible variants.

@Component({...})
export class MyComp {
  // optional
  firstName = input<string>();         // InputSignal<string|undefined>
  age = input(0);                      // InputSignal<number>

  // required
  lastName = input.required<string>(); // InputSignal<string>
}
```

An input is automatically recognized by Angular whenever you use the `input` or `input.required` functions as initializer of class members.

## Why should we use signal inputs and not `@Input()`?

In comparison to decorator-based `@Input`, signal inputs provide numerous benefits:

1. Signal inputs are more **type safe**:
  <br/>• Required inputs do not require initial values, or tricks to tell TypeScript that an input _always_ has a value.
  <br/>• Transforms are automatically checked to match the accepted input values.
2. Signal inputs, when used in templates, will **automatically** mark `OnPush` components as dirty.
3. Values can be easily **derived** whenever an input changes using `computed`.

## Aliasing an input

Angular uses the class member name as the name of the input.
You can alias inputs to change their public name to be different.

```typescript
class StudentDirective {

>   age = input(null, {alias: 'studentAge'});

Author: Done!

}
```

This allows users to bind to your input using `[studentAge]`, while inside your component you can access the input values using `this.age`.

## Using in templates

Signal inputs are non-writable signals.
As with signals declared via `signal()`, you access the current value of the input by calling the input signal.

```html
<p>First name: {{firstName()}}</p>
<p>Last name: {{lastName()}}</p>
```

> This access to the the value is captured in reactive contexts and can notify active consumers, like Angular itself, whenever the input value changes.

Reviewer 2: ```suggestion
This access to the value is captured in reactive contexts and can notify active consumers, like Angular itself, whenever the input value changes.
```

> An input signal in practice is a trivial extension of signals that you know from the [the signals guide](/guide/signals).

Reviewer 2: ```suggestion
An input signal in practice is a trivial extension of signals that you know from [the signals guide](/guide/signals).
```

```typescript
export class InputSignal<T> extends Signal<T> { ... }`.
```

## Deriving values

As with signals, you can derive values from inputs using `computed`.

```typescript
import {Component, input, computed} from '@angular/core';

@Component({...})
export class MyComp {
  age = input(0);

  // age multiplied by two.
  ageMultiplied = computed(() => this.age() * 2);
}
```

Computed signals memoize values.
See more details in the [dedicated section for computed](/guide/signals#computed-signals).

## Monitoring changes

> When using decorator-based inputs (`@Input`), it is difficult to monitor whenever an input's value is changing.

Reviewer: This section comparing with @Input might be a good one to include in the section `Why should we use signal inputs and not @[Input]` To keep focus on `input()` we could start directly with line 110.

Developers used two approaches to monitor value changes:

* using the `ngOnChanges` lifecycle hook.
* using setters to run logic whenever the input changes.

With signal inputs, users can leverage the `effect` function.
The function will execute whenever the input changes.

Consider the following example.
The new value is printed to the console whenever the `firstName` input changes.

```typescript
import {input, effect} from '@angular/core';

class MyComp {
  firstName = input.required<string>();

  constructor() {
    effect(() => {
      console.log(this.firstName());
    });
  }
}
```

> The `fetchUserFromDatabase` function is invoked every time the `firstName` input changes.

Reviewer: ```suggestion
The `console.log` function is invoked every time the `firstName` input changes.
```

# Signal inputs

Signal inputs are an alternative approach to `@Input()`.

<div class="alert is-helpful">

Signal inputs are currently in [developer preview](/guide/releases#developer-preview).

> API might change without deprecation.

Author: Done!

</div>

## What are signal inputs?

Signal inputs are an alternative to decorator-based `@Input()` directive inputs.
They are tightly integrated with the [signals library](/guide/signals) and provide numerous benefits:

1. Signal inputs are more **type safe**:
  <br/>• Required inputs do not require initial values, or tricks to tell TypeScript that an input _always_ has a value.
  <br/>• Transforms are automatically checked to match the accepted input values.
2. Signal inputs, when used in templates, will **automatically** mark `OnPush` components as dirty.
3. Values can be easily **derived** whenever an input changes using `computed`.

## Overview

> Conceptually, signal inputs are similar to the inputs you already know with `@Input`.

Reviewer 3: It is probably obvious but we are not calling out that _values_ are actually expoed as signal. I think it would be helpful to call it out explicitly - otherwise people might get an impression that it more about the authoring that the essence of values as a reactive primitive.

Reviewer: thinking about the audience for this guide, could be Angular devs but also non-Angular devs learning Angular directly with Signals. I think in the overview we should provide a short description that can be understood standalone, without having to go read the guide about `@Input`.

Author: Improving!

> The difference is that you no longer use a decorator to declare an input, but instead the input is automatically recognized by Angular whenever you use the `input` or `input.required` functions as initializer of class members.

Author: Good suggestion. Actually dropping `@input` here completely. Thx

```typescript
import {Component, input} from '@angular/core';

@Component({...})
export class MyComp {
  firstName = input<string>();         // InputSignal<string|undefined>
  lastName = input.required<string>(); // InputSignal<string>

  age = input(0);                      // InputSignal<number>
}
```

## Types of inputs

With signal inputs we have a more clear distinction of possible input variants.
Angular supports:

> * Optional inputs that _can_ have an initial value (see `age` above).

Author: Improved!

* Required inputs that always have a value of the given type.

These two variants are available by the `input` and `input.required` functions exposed by `@angular/core`.

## Using in templates

> Signal inputs are non-writable signals.

Reviewer: ```suggestion
Signal inputs are read-only signals.
```

Reviewer: For readability purposes it's generally advised against negative forms like `non-writable` in documentation. In the signals guide, we say `Signals may be either writable or read-only.`

Author: I don't feel strongly, so happy to change. Do you have any resources on this advise? Note that we were also using `non-writable` in the RFC. See https://github.com/angular/angular/discussions/49683 (and some blog posts also seem to use that). `readonly` works, but doesn't align nicely with the API surface + can be confused easily with the `readonly` modifier of class members.

Reviewer: I think it would be useful to consistently use the same wording across our guides. we refer to computed as read-only as well.

Author: Updated, but I personally wish we would clarify this more somewhere.

As with signals declared via `signal()`, you access the current value of the input by calling the input signal.

> 

Author: Love it!

This access to the the value is captured in reactive contexts and can notify active consumers, like Angular itself, whenever the input value changes.

```html
<p>First name: {{firstName()}}</p>
<p>Last name: {{lastName()}}</p>
```

> An input signal in practice is a trivial extension of signals that were mentioned in [the signals library guide](/guide/signals).

Reviewer 4: ```suggestion
A signal input in practice is a trivial extension of signals that were mentioned in [the signals library guide](/guide/signals).
```

They are called signal inputs instead of input signal over the other parts of the doc

Author: This was an intentional difference to refer to `InputSignal`, but I see what you're saying. Leaving open for others to weigh in.

```typescript
export class InputSignal<T> extends Signal<T> { ... }`.
```

## Deriving values

> Similar to normal signals, you can derive values from inputs using `computed`.

Author: Great suggestion!

Computed signals memoize values.
See more details in the [dedicated section for computed](/guide/signals#computed-signals).

```typescript
import {Component, input, computed} from '@angular/core';

@Component({...})
export class MyComp {
  age = input(0);

  // age multiplied by two.
  ageMultiplied = computed(() => this.age() * 2);
}
```

## Monitoring changes

> Previously, with `@Input` it was rather difficult to monitor whenever an is input changing.

Reviewer 4: ```suggestion
Previously, with `@Input` it was rather difficult to monitor whenever an input is changing.
```

> Previously, with `@Input` it was rather difficult to monitor whenever an is input changing.

Reviewer 5: "to" and "is" are missing in the suggested change.

> When using decorator based inputs (`@Input`) it is difficult **to** monitor whenever an input's value **is** changing.

> Previously, with `@Input` it was rather difficult to monitor whenever an is input changing.

Author: thx updated!

Users monitored inputs using the `ngOnChanges` lifecycle hook, or making use of setters to perform logic whenever the input changes.

With signal inputs, users can make use of the `effect` function to run logic whenever the input changes.

> For example, issuing HTTP requests whenever the first name, or last name inputs change.

Reviewer 4: ```suggestion
For example, issuing HTTP requests whenever the first name input changes.
```

There's no `lastName` input in the example below ?

Author: I was imaging it to be based on the previous sections, but it's likely non-ideal to have that dependency.

> For example, issuing HTTP requests whenever the first name, or last name inputs change.

Author: Updated, but not using http here as per previous commits

```typescript
import {input, effect} from '@angular/core';

class MyComp {
  firstName = input.required<string>();

  constructor() {
    effect(() => {

>       this.fetchUserFromDatabase(this.firstName());

Reviewer 4: Should we include the `onCleanup` function here, to cleanup some subscription maybe, or cancel a promise? Devs will look at this code, and see it as best practice (because it's the docs)

Reviewer 3: Alternatively we should make this example simpler by using say `console.log` - otherwise I agree with @eneajaho that we should make this example more "real life". My vote would go to changing this example to something that doesn't involve http

Author: Using `console.log` to not convolute this example here. Agreed not using http here.

    });
  }
}
```

The `fetchUserFromDatabase` function is invoked every time the `firstName` input changes.
This will happen as soon as `firstName` is available, and for subsequent changes during the lifetime of `MyComp`.

> ## Aliasing an input

Reviewer 3: I would move those sections earlier on as those are more relevant to inputs - the previous discussion in computed / effect is not crucial here (the same logic applies to all signals) and might "distract" from the input essentials.

Author: Sounds good. Moving aliasing.

I did think we want to keep transforms last as `computed` expressions are most of the time the thing people are looking for I'd assume and transforms may be misunderstood then.

> Inputs can be aliased in case the name of the input cannot match the class member name.

Author: reworded

```typescript
class MyComp {
  _rawAge = input(0, {alias: 'age'});
}
```

> With the example above, you will use `this._rawAge` inside your component or template.

Author: I think this is fine, but I'm making the example a bit simpler.

Consumers of `MyComp` can bind to `[age]` using Angular's input binding syntax.

## Value transforms

You may want to coerce or parse input values without changing the meaning of the input.
Transforms convert the raw value from parent templates to the expected input type.
Transforms should be [pure functions](https://en.wikipedia.org/wiki/Pure_function).

<div class="alert is-important">

> Do not use transforms if they change the meaning of the input, or if they are [impure](https://en.wikipedia.org/wiki/Pure_function).

Reviewer 4: ```suggestion
Do not use transforms if they change the meaning of the input, or if they are [impure](https://en.wikipedia.org/wiki/Pure_function#Impure_functions).
```

> Instead, consider using a `computed` for transformations with different meaning, or an `effect` for impure code that should run whenever the input changes.

Reviewer 3: 👍

</div>

> In some cases though, value transforms are a good fit for input.

Reviewer 3: How about inverting things here and showing a "positive" example first, with "avoid" bits after?

Author: Fair suggestion. Not sure. I started with avoid bits beforehand to not encourage transforms too early, but I see that a simpler positive example might actually help. Will check tomorrow

Reviewer: +1 to showing an example of `how/why to use` first and after that share best practices/what to avoid.

Author: Improved! reworked a little more

Consider an input for `disabled` that is accepting `boolean` values.

```typescript
class MyCustomComp {
  disabled = input(false); // InputSignal<boolean>
}
```

Inside your component, you expected `disabled` to be either `true` or `false`.
On the other hand though, users of your component may pass an empty string in templates, as a shorthand, and expect that to be treated as `true`.

```html
<my-custom-comp disabled>
```

This will **fail** type checking because an empty string is not assignable to a `boolean`.
As a component author, you can decide to explicitly support this shorthand by adding a value transform, or by [deriving](#deriving-values) a boolean using `computed`.

```typescript
class MyComp {
  disabled = input(false, {
    transform: (v: boolean|string) => v === '' || v,
  });

  // or alternatively:
  disabledRaw = input<string|boolean>(false, {alias: 'disabled'});

>   disabled = computed(() => this.disabled() === '' || !!this.disabled());

Reviewer: I am confused by the naming here, based on the alias example my understanding is that `this.disabled()` should be `this.disabledRaw()` here.

Author: Good catch!
